nevergrad/functions/games/battleship.py

```diff
@@ -0,0 +1,337 @@
+
+    def __str__(self):
+        serialized = "".join(["".join(b) for b in self.board])
+        print(serialized.count("o"), " alive ships")
+        print(serialized.count("x"), " destroyed ships")
+        print(serialized.count("*"), " failures")
+        stri = "=====================\n"
+        for b in self.board:
+            stri += "".join(b) + "\n"
+        stri += "=====================\n"
+        return stri
+
+    def shoot_consistently(self):
+        candidate_board = self.compliant_board()
+        possible_shoots = []
+        for i in range(self.N):
+            for j in range(self.N):
+                if candidate_board.board[i][j] == "o" and self.board[i][j] != "x":
+                    possible_shoots += [(i, j)]
+        random.seed()
+        s = random.choice(possible_shoots)
+        self.shoot(s[0], s[1])
+
+    def good_shoot(self, policy):
+        random.seed()
+        if policy is not None:
+            index = 0
+            random.seed(17)
+            for i in range(self.N):
+                for j in range(self.N):
+                    if board[i][j] == "o":
+                        index *= random.choice(30) + 1
+                        index += random.choice(range(30))
+                        index = index % 10000000
+                    if board[i][j] == "x":
+                        index *= random.choice(30) + 1
+                        index += random.choice(range(30, 60))
+                        index = index % 10000000
+                    if board[i][j] == "*":
+                        index *= random.choice(30) + 1
+                        index += random.choice(range(60, 90))
+                        index = index % 10000000
+                    if board[i][j] == " ":
+                        index *= random.choice(30) + 1
+                        index += random.choice(range(90, 120))
+                        index = index % 10000000            
+            random.seed(policy[index % len(policy)])
+        possible_shoots = defaultdict(lambda: 0)
+        for _ in range(7):
+            candidate_board = self.compliant_board()
+            #      print "a candidate board for "
+            #      print self
+            #      print " is "
+            #      print candidate_board
+            for i in range(self.N):
+                for j in range(self.N):
+                    if candidate_board.board[i][j] == "o" and self.board[i][j] != "x":
+                        possible_shoots[(i, j)] += 1
+        #    print possible_shoots.keys()
+        max_num = -1
+        assert possible_shoots, str(self) + " has no possible shoots."
+        for s in possible_shoots:
+            if possible_shoots[s] > max_num:
+                max_num = possible_shoots[s]
+                best_shoot = s
+        self.shoot(best_shoot[0], best_shoot[1])
+
+    def compliant_board(self):
+        # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+        # TODO: refactor?
```

Comment:
`# TODO: refactor?` -> I agree :D  
I expect this cant/shouldnt be done now, but such a large function is bound to be buggy somewhere :s